lib/devices/system_reboot.ex

```diff
@@ -0,0 +1,30 @@
+defmodule Onvif.Devices.SystemReboot do
+  import SweetXml
+  import XmlBuilder
+
+  alias Onvif.Device
+
+  def soap_action, do: "http://www.onvif.org/ver10/device/wsdl/SystemReboot"
+
+  @spec request(Device.t()) :: {:ok, any} | {:error, map()}
+  def request(device), do: Onvif.Devices.request(device, __MODULE__)
+
+  def request_body do
+    element(:"s:Body", [element(:"tds:SystemReboot")])
+  end
+
+  def response(xml_response_body) do
+    doc = parse(xml_response_body, namespace_conformant: true, quiet: true)
+
+    parsed_result =
+      xpath(
+        doc,
+        ~x"//s:Envelope/s:Body/tds:SystemRebootResponse"
+        |> add_namespace("s", "http://www.w3.org/2003/05/soap-envelope")
+        |> add_namespace("tds", "http://www.onvif.org/ver10/device/wsdl"),
+        message: ~x"./tds:Message/text()"s
+      )
+
+    {:ok, parsed_result.message}
```

Comment:
is there a failure response we should be exposing with :error?

Reply:
Most of our failure responses are handled by [parse_response](https://github.com/hammeraj/onvif/blob/main/lib/devices/devices.ex#L35) for devices and media interfaces. So the code would short circuit before reaching this line in case of error.

Reply:
Yeah I think those generally trigger on like 400-500s, was unsure if there was a valid xml response with a message like "unable to reboot". docs probably aren't clear on that though

Reply:
Here they have mentioned errors https://www.onvif.org/specs/core/ONVIF-Core-Specification.pdf

```
5.8 Error handling
As with any other protocol, errors can occur during communications, protocol or message processing.
The specification classifies error handling into the following categories:
• Protocol Errors
• SOAP Errors
• Application Errors
```

Even for the SOAP error, I have seen the http status code as 400 only on the axis. Not sure if every manufacturer follows that.

```
{:error,
 %{
   reason: "Received 400 from Elixir.Onvif.Media.Ver10.GetVideoEncoderConfiguration",
   response: "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<SOAP-ENV:Envelope xmlns:SOAP-ENV=\"http://www.w3.org/2003/05/soap-envelope\" xmlns:SOAP-ENC=\"http://www.w3.org/2003/05/soap-encoding\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns:xsd=\"http://www.w3.org/2001/XMLSchema\" xmlns:tns=\"http://tempuri.org/tns.xsd\" xmlns:tns1=\"http://www.onvif.org/ver10/topics\" xmlns:c14n=\"http://www.w3.org/2001/10/xml-exc-c14n#\" xmlns:wsu=\"http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd\" xmlns:ds=\"http://www.w3.org/2000/09/xmldsig#\" xmlns:wsse=\"http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd\" xmlns:wsa=\"http://schemas.xmlsoap.org/ws/2004/08/addressing\" xmlns:wsa5=\"http://www.w3.org/2005/08/addressing\" xmlns:xmime5=\"http://tempuri.org/xmime5.xsd\" xmlns:xop=\"http://www.w3.org/2004/08/xop/include\" xmlns:tplt=\"http://www.onvif.org/ver10/plus/schema\" xmlns:tt=\"http://www.onvif.org/ver10/schema\" xmlns:wstop=\"http://docs.oasis-open.org/wsn/t-1\" xmlns:wsrfr=\"http://docs.oasis-open.org/wsrf/r-2\" xmlns:wsrfbf=\"http://docs.oasis-open.org/wsrf/bf-2\" xmlns:tanae=\"http://www.onvif.org/ver20/analytics/wsdl/AnalyticsEngineBinding\" xmlns:tanre=\"http://www.onvif.org/ver20/analytics/wsdl/RuleEngineBinding\" xmlns:tan=\"http://www.onvif.org/ver20/analytics/wsdl\" xmlns:tds=\"http://www.onvif.org/ver10/device/wsdl\" xmlns:tes-cppb=\"http://www.onvif.org/ver10/events/wsdl/CreatePullPointBinding\" xmlns:tes-e=\"http://www.onvif.org/ver10/events/wsdl/EventBinding\" xmlns:tes-nc=\"http://www.onvif.org/ver10/events/wsdl/NotificationConsumerBinding\" xmlns:tes-np=\"http://www.onvif.org/ver10/events/wsdl/NotificationProducerBinding\" xmlns:tes-ppb=\"http://www.onvif.org/ver10/events/wsdl/PullPointBinding\" xmlns:tes-pps=\"http://www.onvif.org/ver10/events/wsdl/PullPointSubscriptionBinding\" xmlns:tev=\"http://www.onvif.org/ver10/events/wsdl\" xmlns:tes-psmb=\"http://www.onvif.org/ver10/events/wsdl/PausableSubscriptionManagerBinding\" xmlns:wsnt=\"http://docs.oasis-open.org/wsn/b-2\" xmlns:tes-sm=\"http://www.onvif.org/ver10/events/wsdl/SubscriptionManagerBinding\" xmlns:timg=\"http://www.onvif.org/ver20/imaging/wsdl\" xmlns:tmd=\"http://www.onvif.org/ver10/deviceIO/wsdl\" xmlns:tpl=\"http://www.onvif.org/ver10/plus/wsdl\" xmlns:tptz=\"http://www.onvif.org/ver20/ptz/wsdl\" xmlns:tr2=\"http://www.onvif.org/ver20/media/wsdl\" xmlns:trc=\"http://www.onvif.org/ver10/recording/wsdl\" xmlns:trp=\"http://www.onvif.org/ver10/replay/wsdl\" xmlns:trt=\"http://www.onvif.org/ver10/media/wsdl\" xmlns:trv=\"http://www.onvif.org/ver10/receiver/wsdl\" xmlns:tse=\"http://www.onvif.org/ver10/search/wsdl\" xmlns:ter=\"http://www.onvif.org/ver10/error\"><SOAP-ENV:Header></SOAP-ENV:Header><SOAP-ENV:Body><SOAP-ENV:Fault><SOAP-ENV:Code><SOAP-ENV:Value>SOAP-ENV:Sender</SOAP-ENV:Value><SOAP-ENV:Subcode><SOAP-ENV:Value>ter:InvalidArgVal</SOAP-ENV:Value><SOAP-ENV:Subcode><SOAP-ENV:Value>ter:NoConfig</SOAP-ENV:Value></SOAP-ENV:Subcode></SOAP-ENV:Subcode></SOAP-ENV:Code><SOAP-ENV:Reason><SOAP-ENV:Text xml:lang=\"en\">requested VideoEncoderConfiguration does not exist</SOAP-ENV:Text></SOAP-ENV:Reason><SOAP-ENV:Detail></SOAP-ENV:Detail></SOAP-ENV:Fault></SOAP-ENV:Body></SOAP-ENV:Envelope>",
   status: 400
 }}
```

Reply:
5.8.2.2 Generic faults

^ I think these would return as "successes" but we'd have to parse out the failure, like if our user didn't have the permission to reboot the machine.

Reply:
About axis response is also under generic faults `env:Sender ter:InvalidArgVal` and axis send that with 400 status code 🤔

Reply:
Gotcha, I guess we will monitor and address if needed. Thanks for diving in